### PR TITLE
Improve support for data packing

### PR DESCRIPTION
**Pull request by joehoover**
This PR adds support for sequence packing that does not break up sequences. It allows us to take advantage of sequence packing while still strictly respecting input formatting. 

Our current preprocessing implementation supports sequence packing up to a sequence length of `chunk size`. If the final sequence exceeds `chunk size`, it is split so that the packed sequence meets chunk size and the residual tokens are promoted to the beginning of the next sequence. 

The advantage of this implementation is that it produces inputs with a enforced minimum sequence length. However, a disadvantage is that it breaks prompt formats for sequences that have to be split across sequences. 

As an alternative, we've implemented a packing strategy that _does not_ break up sequences. This strategy accordingly cannot enforce a minimum sequence length; however, it does allow us to take advantage of packing while still strictly respecting prompt formats.

### CODE REVIEW

**Comment by joehoover** [comment]
I ran into a few problems that had to be fixed prior to merging. 

First, had to switch from the default data collator to `DataCollatorForDataCollatorForTokenClassification`. Initially, I had switched to `DataCollatorWithPadding`, but I then learned that that collator doesn't pad `labels`. 

Then, it turned out that the token embeddings needed to be resized. In the original llama fine-tuning script, the pad token was added as a special token, but because we were packing sequences into a fixed sequence length, nothing was ever padded. Accordingly, the fact that the token embeddings were missing a dimension was never a problem. 

However, once we started padding the ragged packed sequences, device side asserts were thrown because we were using a token ID that exceeded the vocabulary size of the token embeddings.

File: tests/unit_tests/test_completion_dataset.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    formatted_data = format_data(dataset, tokenizer, dataset_config)
+    assert formatted_data[0]['text'].endswith(tokenizer.eos_token)
+
+@pytest.fixture(scope="session")
+def formatted_dataset(dataset, tokenizer):
+    return format_data(dataset, tokenizer, dataset_config)
+
+def test_tokenize_data_with_wrapped_packing(formatted_dataset, tokenizer, dataset_config):
+    dataset_config.pack_sequences = True
+    dataset_config.wrap_packed_sequences = False
```
Comment (daanelson) [inline tests/unit_tests/test_completion_dataset.py:80]: think this should be `True`

File: train.py [hunk 2]
```diff
@@ -75,9 +75,17 @@ def train(
...
     ),
+    validation_prompt: str = Input(
+        description="Prompt to use for generation during validation. If provided, a response to this prompt will be sampled and logged during validation.",
+        default=None,
+    ),
     learning_rate: float = Input(
         description="learning rate, for learning!", default=1e-4, ge=0
     ),
+    wrap_packed_sequences: bool = Input(
+        description="This is hard to describe succinctly. Maybe we don't need to expose this.",
```
Comment (daanelson) [inline train.py:86]: yeah it's an interesting point, this is hard to describe. I almost wonder (or rather, I do wonder but I don't know for sure) if it makes sense to instead just expose a `pack sequences` flag, and then for `prompt:` `completion:` datasets we don't wrap but for `text:` datasets we do (i.e. users don't get control of this).

...as I think about this, it probably makes the most sense to expose this for now while we run experiments, and then we can re-evaluate once we have our very good excellent killer fine-tunes.

File: llama_recipes/configs/datasets.py [hunk 3]
```diff
@@ -46,4 +46,8 @@ class completion:
     data_path: str = None
     num_validation_samples: int = 100
     run_validation: bool = True
-    validation_data_path: str = None
\ No newline at end of file
+    validation_data_path: str = None
+    pack_sequences: bool = True
```
Comment (daanelson) [inline llama_recipes/configs/datasets.py:50]: imo it'd be nice to expose at least `pack sequences` as an additional argument to `train.py`.

Comment (daanelson) [reply llama_recipes/configs/datasets.py:50]: I tested this out just now and got an error about a nonexistant loss value. Not entirely sure what that's about but also I'm ok tabling it for now (unless some magic solution shows up in the AM).

File: tests/unit_tests/test_completion_dataset.py (hunk 1 above)
Comment (joehoover) [reply tests/unit_tests/test_completion_dataset.py:80]: Good catch!

File: train.py (hunk 2 above)
Comment (joehoover) [reply train.py]: Agreed! I exposed control over packing vs no packing and chunk size, and we can revisit later.

File: llama_recipes/configs/datasets.py (hunk 3 above)
Comment (joehoover) [reply llama_recipes/configs/datasets.py:50]: Hmm, can you share a snippet so I can try to reproduce?

Comment (joehoover) [reply llama_recipes/configs/datasets.py:50]: Fixed. 

Inputs need a `labels` field, but the field was being populated during the packing process. So, when we turned off packing, `labels` were never created. 

I moved the `labels` creation step to the initial tokenization step, which fixed this.
